unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,432 @@
+        let last_leaf = self.flush_buffered_leaf();
+        let root_links = self.flush_buffered_links(NonZeroUsize::new(1).unwrap(), true);
+        // should probably error if there is neither?
+        last_leaf.into_iter().chain(root_links.into_iter())
+    }
+
+    fn flush_buffered_leaf(&mut self) -> Option<(Cid, Vec<u8>)> {
+        if self.block_buffer.is_empty() {
+            return None;
+        }
+        let bytes = self.block_buffer.len();
+
+        let inner = FlatUnixFs {
+            links: Vec::new(),
+            data: UnixFs {
+                Type: UnixFsType::File,
+                Data: Some(Cow::Borrowed(self.block_buffer.as_slice())),
+                filesize: Some(self.block_buffer.len() as u64),
+                // no blocksizes as there are no links
+                ..Default::default()
+            },
+        };
+
+        let (cid, vec) = render_and_hash(inner);
+
+        let total_size = vec.len();
+
+        self.block_buffer.clear();
+
+        // leafs always go to the lowest level
+        if self.unflushed_links.is_empty() {
+            self.unflushed_links.push(Vec::new());
+        }
+
+        self.unflushed_links[0].push((cid.clone(), total_size as u64, bytes as u64));
+
+        Some((cid, vec))
+    }
+
+    // FIXME: collapse the min_links and all into single type to avoid boolean args
+    fn flush_buffered_links(&mut self, min_links: NonZeroUsize, all: bool) -> Vec<(Cid, Vec<u8>)> {
+        /*
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|\   /
+        links#2 |-------------------------------|                         ^^^
+                                                                    one short
+
+        #flush_buffered_links(...) first iterations:
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|==1==|
+        links#2 |-------------------------------|==========================2==|
+
+        new blocks #1 and #2
+
+        #flush_buffered_links(...) last iteration:
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|--1--|
+        links#2 |-------------------------------|--------------------------2--|
+        links#3 |==========================================================3==|
+
+        new block #3 (the root block)
+        */
+        let mut ret = Vec::new();
+
+        for level in 0.. {
+            if !all {
+                if self.unflushed_links[level].len() < min_links.get() {
```

Comment:
are we always safe indexing here?

Reply:
I rewrote this but I am pretty sure I found a panic before that, perhaps with the empty case, so nope, we were not :)